### PR TITLE
Fix Clippy handling of `ExpnKind::Desugaring`

### PR DESCRIPTION
**Pull request by Aaron1011**
PR #72389 exposed a latent issue in Clippy: only the top-most `ExpnKind` was checked for `ExpnKind::Macro` or `ExpnKind::AstPass`. Under certain circumstances, this could result in Clippy emitting warnings for code produced by a macro expansion, due to the topmost `ExpnId` having `ExpnKind::Desugaring`.

This PR tracks the presence of `ExpnKind::Macro` and `ExpnKind::AstPass` in the `call_site` chain through a new `ClosestAstOrMacro`. This enum is used in Clippy, allowing all of Clippy's UI tests to pass. As a result, we can re-enable Clippy tests in `src/bootstrap/test.rs`

### CODE REVIEW

**Comment by Aaron1011** [comment]
cc @flip1995 @nikomatsakis @Manishearth

**Review by Manishearth — APPROVED** [review]
clippy side looks great, thanks!

**Review by flip1995 — APPROVED** [review]

**Comment by flip1995** [comment]
Hm, there is still the `empty_line_after_outer_attribute.rs` clippy test, that fails because of something related to proc_macros (it seems).

Also it seems, that Clippy is not tested on `CI / PR` runs, but only on `CI / auto` runs. I think we should change that in the future, but this is probably something for a different PR.

File: src/librustc_span/hygiene.rs [hunk 1]
```diff
@@ -104,9 +104,19 @@ impl ExpnId {
         HygieneData::with(|data| data.expn_data(self).clone())
     }
 
+    #[inline]
+    pub fn closest_ast_or_macro(self) -> ClosestAstOrMacro {
```
Comment (nikomatsakis) [inline src/librustc_span/hygiene.rs:108]: I think a comment would be good here, even if it just instructs people to visit the `ClosestAstOrMacro` type for more information

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
r=me with nit, but @petrochenkov self-assigned, so probably good to wait for them to weigh in

r? @petrochenkov

**Comment by Aaron1011** [comment]
@flip1995: The `ui/empty_line_after_outer_attribute.rs` test is passing locally for me with this branch.

**Comment by petrochenkov** [comment]
I haven't seen https://github.com/rust-lang/rust/pull/72389 but it appears to be a serious misuse of expansions.
I was pinged on that PR, but unfortunately didn't notice that something named "explain move errors" makes changes to entirely unrelated parts of the compiler.

I need to look at https://github.com/rust-lang/rust/pull/72389 in more detail and see what it does and whether it can be done without hacks.

(That said, `in_external_macro` and friends are also a mess.)

**Comment by petrochenkov** [comment]
I would personally prefer to revert https://github.com/rust-lang/rust/pull/72389 in the meantime and fix clippy.
Otherwise this unexpected task really messes up my work queue and it would be great to delay it until mid-July.

**Comment by Manishearth** [comment]
Yeah like I said before diagnostics should not take priority over existing functionality

**Comment by Aaron1011** [comment]
> I haven't seen #72389 but it appears to be a serious misuse of expansions.

@petrochenkov: Can you elaborate on this? While operators are not 'desugared' in the same way that 'if' statements and 'for' loops are desugared (operators are desugared during MIR building, rather than AST lowering), I believe it fits the way that way that `DesugaringKind` is used. For both 'for' loops and operators, we have non-user-written method calls being inserted, which can be distinguished from explicit calls (`iter.into_iter()`, `foo.add(rhs)`) by examining the `Span`.

**Comment by Manishearth** [comment]
MIR desugaring is not the same as the source-to-source desugaring that occurs in `if let`, MIR is not rust source. DesugaringKind is specifically for stuff that behaves like macros, and that's why it's a part of spans/etc. MIR does _all kinds_ of transforms; closures become structs, etc, at best you should have a different mechanism for tracking that.

When it comes to diagnostics source-to-source transforms are the ones that cause the most pain and that's why they're handled this way.

I hadn't looked closely at this before but now that I look through it I find myself agreeing with @petrochenkov, it's a mistake to lump MIR transforms into DesugaringKind.

**Comment by Aaron1011** [comment]
@Manishearth @petrochenkov: PR https://github.com/rust-lang/rust/pull/73571 is up, which reverts just the operator desugaring changes.

**Comment by petrochenkov** [comment]
Closing in favor of the revert.

**Comment by Aaron1011** [comment]
Now that the revert has landed in https://github.com/rust-lang/rust/pull/73594, I think we should consider landing this. Clippy's current method detecting macro expansions is very fragile - any additional `ExpnKind::Desugaring`s that we apply may trigger the underlying issue that necessitated the revert.

**Comment by Manishearth** [comment]
So I still don't understand how this would lead to a false positive in clippy: could you give an example?

**Comment by Manishearth** [comment]
Like, overall AST desugarings are pretty rare, and it's basically only the initial few that have always existed. Changes to desugarings necessitate changes in Clippy _anyway_ since sometimes we need to un-desugar them.

**Comment by petrochenkov** [comment]
I'm going to close this for now, with https://github.com/rust-lang/rust/pull/72389 reverted it's not super urgent, and I'd like to consider eliminating lowering-time expansions entirely.

**Pull request closed without merging**